gnucash/report/commodity-utilities.scm

```diff
@@ -819,24 +820,26 @@
 ;;    foreign  - foreign commodity/currency
 ;;    domestic - a gnc-monetary pair
 ;;    date     - time64 price
-(define (gnc:case-exchange-time-fn
-         source-option report-currency commodity-list to-date-tp
-         start-percent delta-percent)
+(define* (gnc:case-exchange-time-fn
+          source-option report-currency commodity-list to-date-tp
+          start-percent delta-percent #:key date-hash)
+  (define-syntax-rule (memoize date expensive-fn)
```

Comment:
Its the definition right there. It is a macro.

Reply:
You could copy the whole `gnc:case-exchange-time-fn` definition, and paste into guile repl after `,expand`:

```scheme
cheme@(guile-user)> ,expand (define* (gnc:case-exchange-time-fn
          source-option report-currency commodity-list to-date-tp
          start-percent delta-percent #:key date-hash)
  (define-syntax-rule (memoize date expensive-fn)
    (if date-hash
        (or (hash-ref date-hash date) (hashv-set! date-hash date expensive-fn))
        expensive-fn))
  (case source-option
    ;; Make this the same as gnc:case-exchange-fn
    ((average-cost) (lambda (foreign domestic date)
                      ((memoize date (gnc:make-exchange-function
                                      (gnc:make-exchange-cost-alist
                                       report-currency date)))
                       foreign domestic)))
    ((weighted-average) (lambda (foreign domestic date)
                          (gnc:exchange-by-pricealist-nearest
                           (memoize date (gnc:get-commoditylist-totalavg-prices
                                          commodity-list report-currency date
                                          start-percent delta-percent))
                           foreign domestic date)))
    ((pricedb-before) gnc:exchange-by-pricedb-nearest-before)
    ((pricedb-latest) (lambda (foreign domestic date)
                        (gnc:exchange-by-pricedb-latest foreign domestic)))
    ((pricedb-nearest) gnc:exchange-by-pricedb-nearest)
    (else
     (begin
       (gnc:warn "gnc:case-exchange-time-fn: bad price-source value: "
                 source-option ". Using pricedb-nearest.")
       ;; FIX-ME another hack to prevent report crashing when an
       ;; unimplemented source-option comes through
       gnc:exchange-by-pricedb-nearest))))
$1 = (define* (gnc:case-exchange-time-fn
          source-option
          report-currency
          commodity-list
          to-date-tp
          start-percent
          delta-percent
          #:key
          (date-hash #f #:date-hash))
  (let ((key source-option))
    (cond (((@@ (guile) memv) key '(average-cost))
           (lambda (foreign domestic date)
             ((if date-hash
                (or (hash-ref date-hash date)
                    (hashv-set!
                      date-hash
                      date
                      (gnc:make-exchange-function
                        (gnc:make-exchange-cost-alist
                          report-currency
                          date))))
                (gnc:make-exchange-function
                  (gnc:make-exchange-cost-alist
                    report-currency
                    date)))
              foreign
              domestic)))
          (((@@ (guile) memv) key '(weighted-average))
           (lambda (foreign domestic date)
             (gnc:exchange-by-pricealist-nearest
               (if date-hash
                 (or (hash-ref date-hash date)
                     (hashv-set!
                       date-hash
                       date
                       (gnc:get-commoditylist-totalavg-prices
                         commodity-list
                         report-currency
                         date
                         start-percent
                         delta-percent)))
                 (gnc:get-commoditylist-totalavg-prices
                   commodity-list
                   report-currency
                   date
                   start-percent
                   delta-percent))
               foreign
               domestic
               date)))
          (((@@ (guile) memv) key '(pricedb-before))
           gnc:exchange-by-pricedb-nearest-before)
          (((@@ (guile) memv) key '(pricedb-latest))
           (lambda (foreign domestic date)
             (gnc:exchange-by-pricedb-latest foreign domestic)))
          (((@@ (guile) memv) key '(pricedb-nearest))
           gnc:exchange-by-pricedb-nearest)
          (else
           (gnc:warn
             "gnc:case-exchange-time-fn: bad price-source value: "
             source-option
             ". Using pricedb-nearest.")
           gnc:exchange-by-pricedb-nearest))))
```

Reply:
And what would that do for me?

Reply:
It illustrates why I inserted the macro. To avoid calling expensive-fn unless it's necessary.